File: cmd/goisekai/frontend/lib/reader.js
(() => {
  var root = document.getElementById('reader');
  var canvas = document.getElementById('page-canvas');
  var ctx = canvas.getContext('2d');
  var spinner = document.getElementById('spinner');
  var errPanel = document.getElementById('error-panel');
  var counter = document.getElementById('page-counter');
  var slider = document.getElementById('page-slider');

  // State
  var pid = root.dataset.pluginId,
    mid = root.dataset.mangaId,
    cid = root.dataset.chapterId;
  var pages = [],
    current = 0,
    img = null,
    loading = false,
    _lastFailedRetry = null;
  var viewMode = localStorage.getItem('gi_viewMode') || 'fitWidth';
  var direction = localStorage.getItem('gi_direction') || 'ltr';
  var smoothing = (localStorage.getItem('gi_renderMode') || 'smooth') === 'smooth';
  var zoom = 1,
    panX = 0,
    panY = 0,
    baseScale = 1,
    dpr = 1;
  var preloaded = {}; // page index -> Image
  var nextPages = null; // next chapter page list — fetched lazily for read-ahead spill
  var imgFails = 0; // consecutive image-load failures (at-home nodes flake transiently)
  var readAhead = Math.max(
    0,
    Math.min(10, parseInt(localStorage.getItem('gi_readAhead'), 10) || 3),
  );
  var startPage = parseInt(new URLSearchParams(window.location.search).get('page'), 10);
  var nextChID = root.dataset.nextChapterId || '';
  var prevChID = root.dataset.prevChapterId || '';

  // Auto-hide bars state
  var barsVisible = true;
  var topBar = document.getElementById('top-bar');
  var bottomBar = document.getElementById('bottom-bar');
  var progressLine = document.getElementById('progress-line');
  var zoneCenter = document.getElementById('zone-center');

  function setBarsVisible(visible) {
    barsVisible = visible;
    if (visible) {
      topBar.style.transform = 'translateY(0)';
      bottomBar.style.transform = 'translateY(0)';
      progressLine.style.opacity = '0';
    } else {
      topBar.style.transform = 'translateY(-100%)';
      bottomBar.style.transform = 'translateY(100%)';
      progressLine.style.opacity = '1';
    }
  }

  function updateProgressLine() {
    var total = pages.length || 1;
    var pct = ((current + 1) / total) * 100;
    progressLine.style.width = `${pct}%`;
  }

  // Start with bars hidden
  setBarsVisible(false);

  function imageUrl(p, chapterID) {
    var h = p.headers || {};
    var ref = h.Referer || h.referer || '';
    return (
      '/image?pluginID=' +
      encodeURIComponent(pid) +
      '&url=' +
      encodeURIComponent(p.url) +
      '&mangaID=' +
      encodeURIComponent(mid) +
      '&chapterID=' +
      encodeURIComponent(chapterID || cid) +
      (ref ? `&referer=${encodeURIComponent(ref)}` : '')
    );
  }

  function clampPan() {
    if (!img) return;
    var rect = canvas.getBoundingClientRect();
    var w = img.width * baseScale * zoom,
      h = img.height * baseScale * zoom;
    var cx = Math.max((rect.width - w) / 2, 0);
    var cy = Math.max((rect.height - h) / 2, 0);
    // Image pos = center + pan. Constrain so both edges are reachable.
    var minX = Math.min(0, rect.width - w - cx);
    var maxX = Math.max(0, -cx);
    var minY = Math.min(0, rect.height - h - cy);
    var maxY = Math.max(0, -cy);
    panX = Math.max(minX, Math.min(maxX, panX));
    panY = Math.max(minY, Math.min(maxY, panY));
  }

  function calcBaseScale() {
    if (!img) return;
    var rect = canvas.getBoundingClientRect();
    if (viewMode === 'fitWidth') baseScale = rect.width / img.width;
    else if (viewMode === 'fitHeight') baseScale = rect.height / img.height;
    else baseScale = 1; // original
  }

  function measure() {
    // Canvas must be sized while visible (hard-won lesson): call after paint.
    var rect = canvas.getBoundingClientRect();
    if (rect.width === 0 || rect.height === 0) return false;
    dpr = window.devicePixelRatio || 1;
    canvas.width = Math.round(rect.width * dpr);
    canvas.height = Math.round(rect.height * dpr);
    ctx.setTransform(dpr, 0, 0, dpr, 0, 0);
    calcBaseScale();
    clampPan();
    return true;
  }

  function render() {
    var rect = canvas.getBoundingClientRect();
    ctx.clearRect(0, 0, rect.width, rect.height);
    if (!img) return;
    var w = img.width * baseScale * zoom,
      h = img.height * baseScale * zoom;
    // Center when smaller, top-align when overflowing (top must never clip).
    var x = Math.max((rect.width - w) / 2, 0) + panX;
    var y = Math.max((rect.height - h) / 2, 0) + panY;
    ctx.imageSmoothingEnabled = smoothing;
    ctx.imageSmoothingQuality = 'high';
    ctx.drawImage(img, x, y, w, h);
  }

  function showSpinner(v) {
    spinner.style.display = v ? 'flex' : 'none';
    loading = v;
  }

  function showError(v) {
    errPanel.style.display = v ? 'flex' : 'none';
  }
  function showReaderError(msg) {
    errPanel.querySelector('p').textContent = msg;
    showError(true);
  }
  function hideReaderError() {
    showError(false);
  }

  // ponytail: classify fetch errors into user-friendly text
  function friendlyError(err) {
    if (err instanceof TypeError) return 'Network error — check your connection';
    if (err instanceof SyntaxError) return 'Server returned invalid data';
    var m = err.message || String(err);
    if (/^HTTP \d+$/.test(m)) return 'Server error (' + m.slice(5) + ')';
    return m;
  }

  function showNotice(msg) {
    errPanel.querySelector('p').textContent = msg;
    errPanel.style.display = 'flex';
    setTimeout(() => {
      errPanel.style.display = 'none';
    }, 2000);
  }

  function updateCounter() {
    var total = pages.length || 1;
    counter.textContent = `${current + 1} / ${total}`;
    slider.max = total;
    slider.value = current + 1;
  }

  function reportProgress() {
    var body = new URLSearchParams({
      pluginID: pid,
      mangaID: mid,
      chapterID: cid,
      page: String(current + 1),
    });
    fetch('/action/set-chapter-progress', {
      method: 'POST',
      headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
      body: body.toString(),
    }).catch(() => {});
  }

  function prefetch() {
    // Read-ahead: keep the next K pages warm in the browser cache.
    // Suwayomi-style spill: near the end of the chapter, the leftover
    // budget preloads the first pages of the NEXT chapter instead.
    var remaining = pages.length - 1 - current;
    var budget = Math.min(readAhead, remaining);
    for (let k = 1; k <= budget; k++) {
      const idx = current + k;
      if (!preloaded[idx]) {
        const im = new Image();
        im.src = imageUrl(pages[idx]);
        preloaded[idx] = im;
      }
    }
    var spill = readAhead - budget;
    if (spill > 0 && nextChID) {
      if (nextPages) {
        prefetchNext(spill);
        return;
      }
      // Lazily fetch the next chapter's page list once, then spill into it.
      fetch(`/api/reader-data/${[pid, mid, nextChID].map(encodeURIComponent).join('/')}`)
        .then((r) => r.json())
        .then((d) => {
          nextPages = d?.pages || [];
          if (nextPages.length) prefetchNext(spill);
        })
        .catch(() => {
          /* next-chapter warmup is best-effort */
        });
    }
  }

  function prefetchNext(n) {
    for (let k = 0; k < Math.min(n, nextPages.length); k++) {
      if (!preloaded[`n${k}`]) {
        const im = new Image();
        im.src = imageUrl(nextPages[k], nextChID);
        preloaded[`n${k}`] = im;
      }
    }
  }

  function drawPage(i) {
    current = i;
    showError(false);
    updateCounter();
    updateProgressLine();
    reportProgress();
    prefetch();
    var url = imageUrl(pages[i]);
    var cached = preloaded[i];
    if (cached?.complete && cached.naturalWidth > 0) {
      img = cached;
      afterImage();
      return;
    }
    showSpinner(true);
    var im = new Image();
    im.onload = () => {
      img = im;
      preloaded[i] = im;
      showSpinner(false);
      afterImage();
    };
    im.onerror = () => {
      showSpinner(false);
      imgFails++;
      if (imgFails >= 4) {
        showReaderError('Failed to load image — try Retry');
        return;
      } // give up — manual Retry still available
      // At-home nodes flake transiently: retry the same URL once, then
      // re-fetch the page list (a fresh request may land on a healthy node).
      var reFetch = imgFails % 2 === 0;
      setTimeout(() => {
        if (reFetch) loadChapter(true);
        else drawPage(current);
      }, 1200);
    };
    im.src = url;
  }

  // rAF never fires in a hidden/background tab (Chrome throttles it), which
  // left the first draw pending forever. setTimeout(0) is the safe fallback.
  function nextFrame(fn) {
    var done = false;
    var run = () => {
      if (done) return;
      done = true;
      fn();
    };
    requestAnimationFrame(run);
    setTimeout(run, 50); // hidden tabs never fire rAF — this is the real path
  }

  function afterImage() {
    imgFails = 0;
    if (!measure()) {
      nextFrame(() => {
        measure();
        restoreZoom();
      });
      return;
    }
    restoreZoom();
  }

  function resetZoom() {
    zoom = 1;
    panX = 0;
    panY = 0;
    localStorage.setItem(`gi_zoom_${viewMode}`, '1');
    calcBaseScale();
    render();
  }
  function restoreZoom() {
    zoom = parseFloat(localStorage.getItem(`gi_zoom_${viewMode}`)) || 1;
    panX = 0;
    panY = 0;
    calcBaseScale();
    render();
  }

  function goToPage(i) {
    if (i < 0 || i >= pages.length) return;
    drawPage(i);
  }

  function next() {
    if (current < pages.length - 1) {
      goToPage(current + 1);
      return;
    }
    // Last page → auto-advance to next chapter
    if (nextChID) {
      switchChapter(nextChID, 1);
      return;
    }
    showNotice('Last page');
  }
  function prev() {
    if (current > 0) {
      goToPage(current - 1);
      return;
    }
    // First page → auto-retreat to prev chapter last page
    if (prevChID) {
      switchChapter(prevChID, 'last');
      return;
    }
    showNotice('First page');
  }

  // Fetch-swap navigation: load a chapter in place without a full page reload.
  function switchChapter(targetCID, targetPage) {
    if (loading) return;
    showSpinner(true);
    _lastFailedRetry = function() { switchChapter(targetCID, targetPage); };
    fetch(`/api/reader-data/${[pid, mid, targetCID].map(encodeURIComponent).join('/')}`)
      .then((r) => {
        if (!r.ok) throw new Error(`HTTP ${r.status}`);
        return r.json();
      })
      .then((data) => {
        if (data.error) {
          showReaderError(data.error);
          showSpinner(false);
          return;
        }
        var newPages = data.pages || [];
        if (!newPages.length) {
          showReaderError('Empty chapter (plugin failed to fetch pages) — try Retry');
          showSpinner(false);
          return;
        }
        // Commit new chapter state.
        cid = targetCID;
        pages = newPages;
        nextChID = data.nextChapterID || '';
        prevChID = data.prevChapterID || '';
        preloaded = {};
        nextPages = null;
        imgFails = 0;
        img = null;
        panX = 0;
        panY = 0;
        _lastFailedRetry = null;
        syncChapterNav(data);
        // Update the URL without a reload; back/forward still work.
        var url = `/view/read/${[pid, mid, targetCID].map(encodeURIComponent).join('/')}`;
        history.pushState({}, '', url + (targetPage === 'last' ? '?page=last' : ''));
        var initial = targetPage === 'last' ? pages.length - 1 : 0;
        showSpinner(false);
        drawPage(initial);
      })
      .catch((err) => {
        showReaderError(friendlyError(err));
        showSpinner(false);
      });
  }

  // Reflect chapter state in the top-bar title + prev/next chapter controls.
  function syncChapterNav(data) {
    var t = document.getElementById('chapter-title');
    if (t) {
      const num = data.chapterNum || 0,
        title = data.chapterTitle || '';
      t.textContent = num > 0 ? `Ch. ${num}${title ? ` — ${title}` : ''}` : title;
    }
    var pb = document.getElementById('btn-prev-ch'),
      nb = document.getElementById('btn-next-ch');
    if (pb) {
      pb.style.display = prevChID ? '' : 'none';
      pb.href = prevChID
        ? `/view/read/${[pid, mid, prevChID].map(encodeURIComponent).join('/')}`
        : '#';
    }
    if (nb) {
      nb.style.display = nextChID ? '' : 'none';
      nb.href = nextChID
        ? `/view/read/${[pid, mid, nextChID].map(encodeURIComponent).join('/')}`
        : '#';
    }
  }

  // Zoom controls (cursor-anchored via wheel; buttons zoom about center).
  function zoomBy(factor, cx, cy) {
    var rect = canvas.getBoundingClientRect();
    var mx = cx === undefined ? rect.width / 2 : cx - rect.left;
    var my = cy === undefined ? rect.height / 2 : cy - rect.top;
    var nz = Math.max(0.2, Math.min(5, zoom * factor));
    var sc = nz / zoom;
    panX = mx - (mx - panX) * sc;
    panY = my - (my - panY) * sc;
    zoom = nz;
    localStorage.setItem(`gi_zoom_${viewMode}`, String(zoom));
    clampPan();
    render();
  }

  // Events — wheel on document so it works over click zones + toolbars.
  document.addEventListener(
    'wheel',
    (e) => {
      e.preventDefault();
      if (e.ctrlKey) {
        zoomBy(e.deltaY > 0 ? 0.9 : 1.1, e.clientX, e.clientY);
        return;
      }
      panX -= e.deltaX;
      panY -= e.deltaY;
      clampPan();
      render();
    },
    { passive: false },
  );

  var dragging = false,
    moved = false,
    sx = 0,
    sy = 0;
  canvas.addEventListener('mousedown', (e) => {
    if (e.button !== 0) return;
    dragging = true;
    moved = false;
    sx = e.clientX;
    sy = e.clientY;
    canvas.style.cursor = 'grabbing';
  });
  window.addEventListener('mousemove', (e) => {
    if (!dragging) return;
    var dx = e.clientX - sx,
      dy = e.clientY - sy;
    if (Math.abs(dx) > 5 || Math.abs(dy) > 5) moved = true;
    if (moved) {
      panX += dx;
      panY += dy;
      sx = e.clientX;
      sy = e.clientY;
      clampPan();
      render();
    }
  });
  window.addEventListener('mouseup', () => {
    dragging = false;
    canvas.style.cursor = 'default';
  });
  canvas.addEventListener('dblclick', resetZoom);

  var prevCh = prevChID
    ? `/view/read/${[pid, mid, prevChID].map(encodeURIComponent).join('/')}`
    : null;
  var nextCh = nextChID
    ? `/view/read/${[pid, mid, nextChID].map(encodeURIComponent).join('/')}`
    : null;

  document.getElementById('zone-left').addEventListener('click', () => {
    if (!moved) (direction === 'rtl' ? next : prev)();
  });
  document.getElementById('zone-right').addEventListener('click', () => {
    if (!moved) (direction === 'rtl' ? prev : next)();
  });
  var zt = document.getElementById('zone-top');
  if (prevCh)
    zt.addEventListener('click', () => {
      if (!moved) switchChapter(prevChID, 'last');
    });
  var zb = document.getElementById('zone-bottom');
  if (nextCh)
    zb.addEventListener('click', () => {
      if (!moved) switchChapter(nextChID, 1);
    });
  zoneCenter.addEventListener('click', () => {
    if (!moved) setBarsVisible(!barsVisible);
  });

  // Prev/next chapter buttons in the bottom bar — fetch-swap, not full reload.
  var pb = document.getElementById('btn-prev-ch');
  if (pb)
    pb.addEventListener('click', (e) => {
      e.preventDefault();
      if (prevChID) switchChapter(prevChID, 'last');
    });
  var nb = document.getElementById('btn-next-ch');
  if (nb)
    nb.addEventListener('click', (e) => {
      e.preventDefault();
      if (nextChID) switchChapter(nextChID, 1);
    });

  document.getElementById('btn-prev-page').addEventListener('click', prev);
  document.getElementById('btn-next-page').addEventListener('click', next);
  document.getElementById('btn-retry').addEventListener('click', () => {
    imgFails = 0;
    hideReaderError();
    if (_lastFailedRetry) _lastFailedRetry();
    else loadChapter(true);
  });
  document.getElementById('btn-skip').addEventListener('click', next);
  document.getElementById('btn-zoom-in').addEventListener('click', () => {
    zoomBy(1.2);
  });
  document.getElementById('btn-zoom-out').addEventListener('click', () => {
    zoomBy(1 / 1.2);
  });

  var fitBtn = document.getElementById('btn-fit');
  var fitLabels = { fitWidth: 'Fit W', fitHeight: 'Fit H', original: '1:1' };
  var fitOrder = ['fitWidth', 'fitHeight', 'original'];
  function syncFitLabel() {
    fitBtn.textContent = fitLabels[viewMode];
  }
  fitBtn.addEventListener('click', () => {
    viewMode = fitOrder[(fitOrder.indexOf(viewMode) + 1) % fitOrder.length];
    localStorage.setItem('gi_viewMode', viewMode);
    syncFitLabel();
    restoreZoom();
  });
  syncFitLabel();

  var dirBtn = document.getElementById('btn-dir');
  function syncDirLabel() {
    dirBtn.textContent = direction.toUpperCase();
  }
  dirBtn.addEventListener('click', () => {
    direction = direction === 'ltr' ? 'rtl' : 'ltr';
    localStorage.setItem('gi_direction', direction);
    syncDirLabel();
  });
  syncDirLabel();

  slider.addEventListener('input', () => {
    goToPage(parseInt(slider.value, 10) - 1);
  });

  document.addEventListener('keydown', (e) => {
    if (e.target.matches('input, select, textarea')) return;
    if (e.key === 'ArrowRight' || e.key === 'd') {
      e.preventDefault();
      (direction === 'rtl' ? prev : next)();
    } else if (e.key === 'ArrowLeft' || e.key === 'a') {
      e.preventDefault();
      (direction === 'rtl' ? next : prev)();
    } else if (e.key === ' ') {
      e.preventDefault();
      next();
    } else if (e.key === 'Escape') {
      e.preventDefault();
      window.location.href = `/view/manga/${encodeURIComponent(pid)}/${encodeURIComponent(mid)}`;
    } else if (e.key === 'Home') {
      e.preventDefault();
      goToPage(0);
    } else if (e.key === 'End') {
      e.preventDefault();
      goToPage(pages.length - 1);
    } else if (e.key === 'r' || e.key === 'R') {
      e.preventDefault();
      resetZoom();
    }
  });

  // Resize: recompute base scale only — keep user zoom/pan.
  window.addEventListener('resize', () => {
    if (measure()) render();
  });

  // Back/forward: the URL changed via pushState — reload that chapter in place.
  window.addEventListener('popstate', () => {
    var parts = location.pathname.split('/');
    // /view/read/{pluginID}/{mangaID}/{chapterID} → last segment is the chapter ID.
    var target = decodeURIComponent(parts[parts.length - 1] || '');
    if (target && target !== cid) {
      const page = new URLSearchParams(location.search).get('page');
      switchChapter(target, page === 'last' ? 'last' : 1);
    }
  });

  // Boot: fetch page data, then first draw after the canvas is visible.
  function loadChapter(resume) {
    preloaded = {}; // stale at-home URLs 404 — drop cached Image objects
    _lastFailedRetry = function() { loadChapter(resume); };
    fetch(`/api/reader-data/${[pid, mid, cid].map(encodeURIComponent).join('/')}`)
      .then((r) => {
        if (!r.ok) throw new Error(`HTTP ${r.status}`);
        return r.json();
      })
      .then((data) => {
        if (data.error) {
          showReaderError(data.error);
          return;
        }
        pages = data.pages || [];
        if (!pages.length) {
          showReaderError('Empty chapter (plugin failed to fetch pages) — try Retry');
          return;
        }
        _lastFailedRetry = null;
        nextFrame(() => {
          // Honor ?page=N (or ?page=last from chapter advance); a retry keeps
          // the page the reader was already on.
          var sp = resume ? current + 1 : startPage;
          if (window.location.search.indexOf('page=last') >= 0) sp = pages.length;
          var initial = sp ? Math.max(1, Math.min(sp, pages.length)) - 1 : 0;
          drawPage(initial);
        });
      })
      .catch((err) => {
        showReaderError(friendlyError(err));
      });
  }
  loadChapter();
})();

File: internal/templates/views/reader.lua
-- views/reader.lua
-- Reader shell (blank layout). Replaces views/reader.jet.
-- Called as: reader(data) -> string (body HTML only, blank layout wraps it)
-- data keys: PluginID, MangaID, Manga (types.Manga), ChapterID,
--   Chapters, CurrentChapter (types.Chapter), PrevChapterID, NextChapterID

return function(data)
	local pluginID = data.PluginID or ""
	local mangaID = data.MangaID or ""
	local chapterID = data.ChapterID or ""
	local manga = data.Manga or {}
	local mangaTitle = manga.Title or ""
	local currentChapter = data.CurrentChapter or {}
	local chNum = currentChapter.ChapterNum or 0
	local chTitle = currentChapter.Title or ""
	local prevChapterID = data.PrevChapterID or ""
	local nextChapterID = data.NextChapterID or ""

	local parts = {}
	local function emit(s)
		parts[#parts + 1] = s
	end

	-- Reader container with data-* attributes for reader.js
	emit('<div id="reader" class="relative w-full h-screen"')
	emit('  data-plugin-id="' .. h(pluginID) .. '"')
	emit('  data-manga-id="' .. h(mangaID) .. '"')
	emit('  data-chapter-id="' .. h(chapterID) .. '"')
	emit('  data-manga-title="' .. h(mangaTitle) .. '"')
	emit('  data-next-chapter-id="' .. h(nextChapterID) .. '"')
	emit('  data-prev-chapter-id="' .. h(prevChapterID) .. '">')

	-- Top bar
	emit("  <!-- Top bar: persistent (back + page counter + controls) -->")
	emit(
		'  <div id="top-bar" class="absolute top-0 left-0 right-0 z-30 flex flex-wrap items-center gap-2 bg-neutral-950/85 backdrop-blur-sm px-3 py-2 text-sm transition-transform duration-300 ease-in-out">'
	)
	emit(
		'  <a href="/view/manga/'
			.. h(pluginID)
			.. "/"
			.. h(mangaID)
			.. '" class="border border-neutral-700 hover:bg-neutral-800 active:scale-95 transition rounded-md px-3 py-1.5">← Back</a>'
	)
	emit('  <span class="ml-2 text-xs cursor-default" title="← → navigate · Space next · Esc back">⌨</span>')
	emit('    <span class="text-neutral-500 truncate max-w-[25%]">' .. h(mangaTitle) .. "</span>")
	emit('    <span id="page-counter" class="text-neutral-400">– / –</span>')
	emit('    <div class="flex-1 flex justify-center min-w-0">')
	emit('      <span id="chapter-title" class="text-neutral-300 truncate">')
	if chNum > 0 then
		emit("        Ch. " .. h(tostring(chNum)))
		if chTitle ~= "" then
			emit(" — " .. h(chTitle))
		end
	else
		emit("        " .. h(chTitle))
	end
	emit("      </span>")
	emit("    </div>")
	emit('    <div class="flex items-center gap-1">')
	emit(
		'      <button id="btn-fit" title="Fit mode" class="border border-neutral-700 hover:bg-neutral-800 active:scale-95 transition rounded-md px-3 py-1.5 text-xs">Fit W</button>'
	)
	emit(
		'      <button id="btn-zoom-out" title="Zoom out" class="border border-neutral-700 hover:bg-neutral-800 active:scale-95 transition rounded-md px-3 py-1.5 text-xs">−</button>'
	)
	emit(
		'      <button id="btn-zoom-in" title="Zoom in" class="border border-neutral-700 hover:bg-neutral-800 active:scale-95 transition rounded-md px-3 py-1.5 text-xs">＋</button>'
	)
	emit(
		'      <button id="btn-dir" title="Direction (RTL/LTR)" class="border border-neutral-700 hover:bg-neutral-800 active:scale-95 transition rounded-md px-3 py-1.5 text-xs">LTR</button>'
	)
	emit("    </div>")
	emit("  </div>")

	-- Click zones
	emit("  <!-- Click zones: left/right = page, top/bottom strip = chapter -->")
	emit(
		'  <div id="zone-top" class="absolute top-12 left-[35%] right-[35%] h-[18%] z-10 cursor-pointer" title="Previous chapter"></div>'
	)
	emit(
		'  <div id="zone-bottom" class="absolute bottom-14 left-[35%] right-[35%] h-[18%] z-10 cursor-pointer" title="Next chapter"></div>'
	)
	emit(
		'  <div id="zone-left" class="absolute top-12 bottom-14 left-0 w-[35%] z-10 cursor-pointer" title="Previous page"></div>'
	)
	emit(
		'  <div id="zone-right" class="absolute top-12 bottom-14 right-0 w-[35%] z-10 cursor-pointer" title="Next page"></div>'
	)

	-- Canvas
	emit("  <!-- Canvas -->")
	emit('  <canvas id="page-canvas" class="absolute inset-0 h-full w-full"></canvas>')

	-- Center click zone
	emit("  <!-- Center click zone: toggles bars visibility (before side zones in DOM so nav strips win overlaps) -->")
	emit(
		'  <div id="zone-center" class="absolute top-12 bottom-14 left-[35%] right-[35%] z-10 cursor-pointer" title="Toggle bars"></div>'
	)

	-- Spinner
	emit("  <!-- Spinner -->")
	emit('  <div id="spinner" class="absolute inset-0 z-20 hidden items-center justify-center bg-neutral-950/70">')
	emit('    <div class="h-10 w-10 animate-spin rounded-full border-2 border-neutral-700 border-t-indigo-500"></div>')
	emit("  </div>")

	-- Error panel
	emit(
		'  <!-- Error card: centered between toolbars, non-blocking. Inline display:none keeps the "Failed to load page" flash invisible until the stylesheet applies (Tailwind `hidden` alone leaves it visible pre-CSS). -->'
	)
	emit(
		'  <div id="error-panel" style="display:none" class="absolute top-12 bottom-14 left-0 right-0 z-40 hidden items-center justify-center pointer-events-none">'
	)
	emit(
		'    <div class="bg-black/85 backdrop-blur-sm rounded-lg p-6 max-w-md w-[90%] flex flex-col items-center gap-3 pointer-events-auto">'
	)
	emit('      <p class="text-sm text-red-400">Failed to load page</p>')
	emit('      <div class="flex gap-2">')
	emit(
		'        <button id="btn-retry" class="bg-indigo-600 hover:bg-indigo-500 rounded-md px-4 py-2 text-sm">↻ Retry</button>'
	)
	emit(
		'        <button id="btn-skip" class="border border-neutral-700 hover:bg-neutral-800 rounded-md px-4 py-2 text-sm">→ Skip</button>'
	)
	emit("      </div>")
	emit("    </div>")
	emit("  </div>")

	-- Bottom bar
	emit("  <!-- Bottom bar: chapter nav + page slider -->")
	emit(
		'  <div id="bottom-bar" class="absolute bottom-0 left-0 right-0 z-30 flex items-center gap-3 bg-neutral-950/85 backdrop-blur-sm px-3 py-2 text-sm transition-transform duration-300 ease-in-out">'
	)
	local prevStyle = prevChapterID == "" and ' style="display:none"' or ""
	emit(
		'    <a id="btn-prev-ch" href="/view/read/'
			.. h(pluginID)
			.. "/"
			.. h(mangaID)
			.. "/"
			.. h(prevChapterID)
			.. '" class="border border-neutral-700 hover:bg-neutral-800 active:scale-95 transition rounded-md px-3 py-1.5"'
			.. prevStyle
			.. ">← Prev ch.</a>"
	)
	emit(
		'    <button id="btn-prev-page" class="border border-neutral-700 hover:bg-neutral-800 active:scale-95 transition rounded-md px-3 py-1.5">‹</button>'
	)
	emit('    <input id="page-slider" type="range" min="1" value="1" class="h-2 flex-1 accent-indigo-500" />')
	emit(
		'    <button id="btn-next-page" class="border border-neutral-700 hover:bg-neutral-800 active:scale-95 transition rounded-md px-3 py-1.5">›</button>'
	)
	local nextStyle = nextChapterID == "" and ' style="display:none"' or ""
	emit(
		'    <a id="btn-next-ch" href="/view/read/'
			.. h(pluginID)
			.. "/"
			.. h(mangaID)
			.. "/"
			.. h(nextChapterID)
			.. '" class="border border-neutral-700 hover:bg-neutral-800 active:scale-95 transition rounded-md px-3 py-1.5"'
			.. nextStyle
			.. ">Next ch. →</a>"
	)
	emit("  </div>")

	-- Progress line
	emit("  <!-- Progress line: visible when bars are hidden -->")
	emit(
		'  <div id="progress-line" class="absolute bottom-0 left-0 right-0 z-40 h-2 bg-indigo-500 shadow-[0_0_8px_2px_rgba(99,102,241,0.55)] transition-opacity duration-300 ease-in-out opacity-0"></div>'
	)

	emit("</div>")

	-- reader.js
	emit('<script src="/static/lib/reader.js" defer></script>')
	emit("<script>")
	emit("(function() {")
	emit("  var title = '" .. h(mangaTitle) .. "';")
	emit("  var chNum = " .. tostring(chNum) .. ";")
	emit("  var chTitle = '" .. h(chTitle) .. "';")
	emit("  if (title) {")
	emit("    var label = title;")
	emit("    if (chNum > 0) { label += ' \\u2013 Ch. ' + chNum; if (chTitle) label += ': ' + chTitle; }")
	emit("    else if (chTitle) { label += ' \\u2013 ' + chTitle; }")
	emit("    document.title = label + ' \\u2013 goIsekai';")
	emit("  }")
	emit("")
	emit("  var dirBtn = document.getElementById('btn-dir');")
	emit("  if (dirBtn) {")
	emit("    var RTL_CLASSES = 'bg-indigo-600/20 border-indigo-500/40';")
	emit("    function applyDirStyle() {")
	emit("      if (localStorage.getItem('gi_direction') === 'rtl') dirBtn.classList.add(...RTL_CLASSES.split(' '));")
	emit("      else dirBtn.classList.remove(...RTL_CLASSES.split(' '));")
	emit("    }")
	emit("    applyDirStyle();")
	emit("    dirBtn.addEventListener('click', function() {")
	emit("      requestAnimationFrame(applyDirStyle);")
	emit("    });")
	emit("  }")
	emit("})();")
	emit("</script>")

	return table.concat(parts, "\n")
end
